### ISSUE TITLE
How to reproduce?

### ISSUE THREAD

**Issue by EoinKenny** [ximingwen/GAProtoNet#1]
Hi is there supposed to be a train.py file in src/ ?

I would like to use your prototype classifier, but how to do this? A minimal working examples on IMDB would be ok!

**Comment by ximingwen** [comment]
Hi,

Thank you very much for your question!  We do have a train.py file within src/ with the command in Readme and you can download the IMDb datasets with the link provided in our paper.

**Comment by EoinKenny** [comment]
Thanks for the help, I think your requirements requires a bit of fine-tuning, 

Doesn't install pytorch correctly on MacOS
pandas isn't there
pip install tensorboardX
sklearn
etc...

```
XLNetTokenizer requires the SentencePiece library but it was not found in your environment. Checkout the instructions on the
installation page of its repo: https://github.com/google/sentencepiece#installation and follow the ones
that match your environment. Please note that you may need to restart your runtime after installation.
```

I'm struggling a bit to get a script going, can you by any chance try on your end to make the env from scratch and run one training process on IMDB?

Thank you so much!

**Comment by ximingwen** [comment]
Hi Eoin!

We updated the readme and requirement.txt so it has a more detailed description of how to replicate our work. Could you please have a try again? Let me know if you have any question!

**Comment by EoinKenny** [comment]
Hi!

I was able to get a version of this code running thanks, but I have a question

```
        self.encoder_key = [nn.Linear(in_features=prototype_dim, out_features=k_dim) for i in range(num_heads)]
        self.encoder_query = [nn.Linear(in_features=bert_cls_dim, out_features=q_dim) for i in range(num_heads)]
        self.encoder_value = [nn.Linear(in_features=prototype_dim, out_features=v_dim) for i in range(num_heads)]
        self.fc = nn.Linear(in_features=num_heads*v_dim, out_features=v_dim)
        
    def forward(self, input_ids, attention_mask, token_type_ids, labels):
        cls_token, __, __  = self.llm( input_ids, attention_mask, token_type_ids, labels)
        multi_graph_encoded_cls = []
               
        for i in range(self.num_heads):
            # encoded q, k ,v
            encoded_queries = self.encoder_query[i].to(self.device)(cls_token)
            encoded_proto_keys  = self.encoder_key[i].to(self.device)(self.prototype_layer.prototype_vectors)
            encoded_proto_values = self.encoder_value[i].to(self.device)(self.prototype_layer.prototype_vectors)

```

This is where the prototypes are used in the code, but when I look at the forward pass in this part

```
    def forward(self, x, labels):

        encoded_key = self.encoder_key(self.prototype_vectors)
        encoded_value = self.encoder_value(self.prototype_vectors)
        # encoded_key = self.prototype_vectors
        # print(self.prototype_vectors.shape)
        # print(encoded_key.shape)
        # z = self.linear(x)
        # cosine_similarities = F.cosine_similarity(z.unsqueeze(1), self.prototype_vectors.unsqueeze(0), dim=-1)
        # token_sequences = cosine_similarities.unsqueeze(2) * self.prototype_vectors.unsqueeze(0)
        # loss = self.loss_weights['diversity_loss_z_p'] * self.diversity_loss_z_p(z) + self.loss_weights['diversity_loss_p_z'] * self.diversity_loss_p_z(z) + self.loss_weights['diversity_loss_p_p'] * self.diversity_loss_p_p() + self.loss_weights['loss_num_prototypes'] * self.loss_num_prototypes(z) + self.loss_weights['cluster_loss'] * self.cluster_loss(z, labels, cosine_similarities) + self.loss_weights['seperation_loss'] * self.cluster_loss(z, labels, cosine_similarities)
        # return token_sequences, loss

        return encoded_key, encoded_value

```

It looks incomplete.

Like, what I mean is that I don't know how you could project these prototypes to real training examples. The prototypes are a different dimension to e.g. BERT and don't seem to come from the CLS space.

For example if I set my prototype dim=128, how exactly do I input the training data to find the nearest neighbor to this? It seems it's the self.linear above, but you have commented it out.

Could you please explain how to input some text example in the network, and measure it's L2 distance to the prototypes? Thank you!!! :-) 

The input data is mapped to the CLS here

```
    def forward(self, input_ids, attention_mask, token_type_ids, labels):
        cls_token, __, __  = self.llm( input_ids, attention_mask, token_type_ids, labels)
        multi_graph_encoded_cls = []
               
        for i in range(self.num_heads):
            # encoded q, k ,v
            encoded_queries = self.encoder_query[i].to(self.device)(cls_token)
            encoded_proto_keys  = self.encoder_key[i].to(self.device)(self.prototype_layer.prototype_vectors)
            encoded_proto_values = self.encoder_value[i].to(self.device)(self.prototype_layer.prototype_vectors)

```

But that's a different dimension to the prototypes, it seem (again) like you're supposed to put it through self.linear in the PrototypeLayer class, but it's commented out
